### ISSUE TITLE
Can't destroy parent if not paranoid

### ISSUE THREAD

**Issue by szechyjs** [ActsAsParanoid/acts_as_paranoid#151]
If a parent model isn't paranoid it's impossible to destroy it. It calls `#destroy` on all the children, which doesn't actually delete them, then when it goes to delete the parent, it gets a foreign key violation from the database.

```rb
class Parent < ActiveRecord::Base
  has_many :children, dependent: :destroy
end

class Child < ActiveRecord::Base
  acts_as_paranoid

  belongs_to :parent
end

Parent.first.destroy

ActiveRecord::InvalidForeignKey: PG::ForeignKeyViolation: ERROR:  update or delete on table "parents" violates foreign key constraint "fk_rails_b38590a748" on table "children"
DETAIL:  Key (id)=(aa67c98c-d81f-5a9c-b0bc-26caa0051aea) is still referenced from table "children".
```

Since `Parent` is not paranoid I would expect calling `#destroy` to work and fully destroy its children.

If I add `acts_as_paranoid` to the parent it works just fine, however I have no need to soft delete the parent.

_Originally posted by @szechyjs in https://github.com/ActsAsParanoid/acts_as_paranoid/issues/142#issuecomment-589810918_

**Comment by mvz** [comment]
This behavior is as intended. ActsAsParanoid aims to never allow accidental full destruction of a record. It cannot know that the destruction of Child objects is fine for you in this case.

The simplest solution for you would be to explicitly destroy the chidren with `#destroy_fully` before calling `#destroy` on the parent. This requires no change to ActsAsParanoid.

Going a bit further, I think there are three behaviors involving deletion of the Child that would be possible for a `has_many`/`belongs_to` relation (I'm leaving out the other options like nullify and restrict_with_error):

1. Fully destroy all children. This is what you're trying to achieve with `dependent: :destroy`.
2. Soft-delete all children. This is what actually happens with `dependent: :destroy`, but it's broken for your case since the foreign key is not nullified.
3. Delete the children from the database without any callbacks. This happens with `dependent: :delete_all` and may work for you depending on other factors.

It may be possible to add a mechanism to make case 1 work by adding some `dependent: :destroy_fully` option. Changing case 2 so that it works for all use cases is worth investigating but seems tricky.

**Comment by szechyjs** [comment]
I agree that case 1 would probably be the best fit here, as it would be opt-in (`dependent: :destroy_fully`).

**Comment by diegotc86** [comment]
Hi, I've been struggling with exactly this problem for a while. I finally solved it this way:
```ruby
# Parent model
before_destroy do
    children.with_deleted.each(&:destroy_fully!)
end
```
You can leave the option `dependent: :destroy` or remove it. If you leave it, it first will make a soft delete on every child and then the before_destroy block will delete permanently the children.

Hope this is helpful for someone out there.

**Comment by Atashia-Sameen** [comment]
I have a scenario where the grand parent model is paranoid and the parent is not paranoid and the child is again paranoid.

class GrandParent
  acts_as_paranoid 
  has_many :parents, dependent: :destroy
end

class Parent
 belongs_to :grand_parent
   has_many :children,
           -> { with_deleted },
           inverse_of: :parent,
           dependent: :nullify
end

class Child
 acts_as_paranoid
   belongs_to :parent, inverse_of: :childs, optional: true
end

I am getting the following error while deleting the Grandparant

ActiveRecord::InvalidForeignKey (PG::ForeignKeyViolation: ERROR:  update or delete on table "parent" violates foreign key constraint "fk_rails_b10c80c09a" on table "child"
DETAIL:  Key (id)=(35) is still referenced from table "child".
):

**Comment by mvz** [comment]
Hi @Atashia-Sameen that looks like a different problem since you're using `dependent: :nullify` inside `Parent`, instead of `dependent: :destroy`. Could you please open a new issue for this?
